overlord/devicestate/devicemgr.go

```diff
@@ -2214,3 +2214,140 @@ func (m *DeviceManager) RemoveRecoveryKeys() error {
 
 	return secbootRemoveRecoveryKeys(recoveryKeyDevices)
 }
+
+// EncryptionSupportInfo describes what encryption is available and needed
+// for the current device
+type EncryptionSupportInfo struct {
+	// Disabled is set if to true encryption was furcefully
+	// disabled (e.g. via the seed pa partition), if set the rest
+	// of the struct content is not relevant.
+	Disabled bool
+
+	// StorageSafety describes the level safety properties
+	// requested by the model
+	StorageSafety asserts.StorageSafety
+	// Available is set to true if encryption is available on this device
+	// with the used gadget.
+	Available bool
+
+	// Type us set to the EncryptionType that can be used if
+	// Available is true.
+	Type secboot.EncryptionType
+
+	// UnvailableErr is set if the encryption support availability of
+	// the this device and used gadget do not match the
+	// storage safety requirements.
+	UnavailableErr error
+	// UnavailbleWarning describes why encryption support is not
+	// available in case it is optional.
+	UnavailableWarning string
+}
+
+var secbootCheckTPMKeySealingSupported = secboot.CheckTPMKeySealingSupported
+
+// checkEncryption verifies whether encryption should be used based on the
+// model grade and the availability of a TPM device or a fde-setup hook
+// in the kernel.
+func (m *DeviceManager) checkEncryption(st *state.State, deviceCtx snapstate.DeviceContext) (secboot.EncryptionType, error) {
+	model := deviceCtx.Model()
+
+	kernelInfo, err := snapstate.KernelInfo(st, deviceCtx)
+	if err != nil {
+		return "", fmt.Errorf("cannot check encryption support: %v", err)
+	}
+	var gadgetInfo *gadget.Info
+	if gadgetSnapInfo, err := snapstate.GadgetInfo(st, deviceCtx); err == nil {
```

Comment:
nitpick: i'd encapsulate this (2259-2264) in a lambda or seperate function called `maybeReadGadgetInfo`, it would make the code more readable and clarify intent

Reply:
Thank you, great point. I actually updated the tests now so that we have a gadget everywhere needed and that makes the code less confusing as we can now follow the standard "if err != nil" pattern. Sorry that I overlooked this earlier and thanks for pointing it out!